### PR TITLE
R4R: Censorship Slash

### PR DESCRIPTION
**Pull request by MrXJC**
Target Issue: https://github.com/irisnet/irishub/issues/1052

### CODE REVIEW

File: modules/slashing/keeper.go [hunk 1]
```diff
@@ -174,6 +174,50 @@ func (k Keeper) handleValidatorSignature(ctx sdk.Context, addr crypto.Address, p
 	return
 }
 
+// Punish proposer censorship by slashing malefactor's stake
+func (k Keeper) handleProposalCensorship(ctx sdk.Context, addr crypto.Address, infractionHeight int64) (tags sdk.Tags) {
```
Comment (wukongcheng) [inline modules/slashing/keeper.go]: handleProposalCensorship => handleProposerCensorship

File: modules/slashing/tick.go [hunk 2]
```diff
@@ -40,3 +40,20 @@ func BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock, sk Keeper) (tags
 
 	return
 }
+
+// slashing begin block functionality
```
Comment (wukongcheng) [inline modules/slashing/tick.go]: slashing end block functionality

**Comment by wukongcheng** [comment]
If a Validator is slash twice at the same block height(double-sign & invalid tx), each time slash 20%, then the total slash is 36% instead of 40%.

File: modules/slashing/keeper.go [hunk 3]
```diff
@@ -174,6 +174,50 @@ func (k Keeper) handleValidatorSignature(ctx sdk.Context, addr crypto.Address, p
...
+	logger.Info(fmt.Sprintf("proposer censorship from %s at height %d", pubkey.Address(), infractionHeight))
+
+
+	// Slash validator
+	// `power` is the int64 power of the validator as provided to/by
+	// Tendermint. This value is validator.Tokens as sent to Tendermint via
+	// ABCI, and now received as evidence.
+	// The revisedFraction (which is the new fraction to be slashed) is passed
+	// in separately to separately slash unbonding and rebonding delegations.
+	tags = k.validatorSet.Slash(ctx, consAddr,infractionHeight, validator.GetPower().RoundInt64(), sdk.NewDecWithPrec(5,2))
```
[the remark this answers is not in the document]
Comment (haifengxi) [reply modules/slashing/keeper.go]: Any new param or param change needs to be added to the params spec. @MrXJC @wukongcheng

[the remark this answers is not in the document]
Comment (MrXJC) [reply modules/slashing/keeper.go]: OK we also need a ProposerCensorshipUnbondDuration like DoubleSignUnbondDuration

**Pull request merged**